feat: power dashboard

rahulguptajss opened the pull request:
![image](https://user-images.githubusercontent.com/25551691/160659305-8c2538da-20cd-414d-af7a-50205b7f8893.png)
![image](https://user-images.githubusercontent.com/25551691/160659350-80ca6a8c-46a4-4468-b7a9-740b7be50a75.png)

cgrinds commented on cmd/collectors/zapi/plugins/shelf/shelf.go:
```diff
@@ -226,12 +231,18 @@ func (my *Shelf) calculateEnvironmentMetrics(output []*matrix.Matrix, data *matr
 			m := data.GetMetric(k)
 			instance := data.GetInstance(key)
+			if instance == nil {
+				my.Logger.Warn().Str("key", key).Msg("Instance not found")
```
should we continue or fall through when instance is nil?

cgrinds commented on grafana/dashboards/cmode/harvest_dashboard_power.json:
```diff
@@ -0,0 +1,2185 @@
+    ]
+  },
+  "timezone": "",
+  "title": "NetApp Detail: Power Dashboard",
```
`NetApp Detail: Power` instead of `NetApp Detail: Power Dashboard`

cgrinds commented on grafana/dashboards/cmode/harvest_dashboard_power.json:
```diff
@@ -0,0 +1,2185 @@
+      },
+      "id": 78,
+      "panels": [],
+      "title": "Shelve",
```
Shelves

cgrinds commented on grafana/dashboards/cmode/harvest_dashboard_power.json:
```diff
@@ -0,0 +1,2185 @@
+      },
+      "id": 76,
+      "panels": [],
+      "title": "Node",
```
Nodes

cgrinds reviewed (commented):
nice!

cgrinds commented:
General comments
* Dashboard panel queries do not work with variable `Cluster: All`
* Resize the bottom two panels to 50%, 50% so they take the full horizontal space
* For new dashboards, we should probably use the “Time Series” panel instead of the “Graph (old)” panel – in lots of cases, edit, migrate works to convert
* Only one digit of significance for temperature
* The lower shelf table, it’s unfortunate you need to scroll right so much to see the most relevant info – can we move some of the left-most columns after the status or hide some of them by default?
* Can we right justify numbers in all of the tables?

rahulguptajss commented:
> General comments
> 
> * Dashboard panel queries do not work with variable `Cluster: All`
> * Resize the bottom two panels to 50%, 50% so they take the full horizontal space
> * For new dashboards, we should probably use the “Time Series” panel instead of the “Graph (old)” panel – in lots of cases, edit, migrate works to convert
> * Only one digit of significance for temperature
> * The lower shelf table, it’s unfortunate you need to scroll right so much to see the most relevant info – can we move some of the left-most columns after the status or hide some of them by default?
> * Can we right justify numbers in all of the tables?

- `Cluster: All` is working fine.
- Resize done
- Changed to `Time Series` Panel
- One digit of significance now for temp
- Moved columns around
- Tested against one cluster. We'll need to find more clusters having power information.

rahulguptajss commented:
![image](https://user-images.githubusercontent.com/25551691/160771968-512fd315-49ab-4d70-a692-ed47084904a3.png)
![image](https://user-images.githubusercontent.com/25551691/160772002-4d4567e3-12ef-4c7f-956c-b57c01f5aad1.png)

Hardikl commented on grafana/dashboards/cmode/harvest_dashboard_power.json:
```diff
@@ -0,0 +1,2105 @@
+  "timezone": "",
+  "title": "NetApp Detail: Power",
+  "uid": "iY7zXdynk",
+  "version": 97
```
1

Hardikl commented on grafana/dashboards/cmode/harvest_dashboard_power.json:
```diff
@@ -0,0 +1,2105 @@
+    ]
+  },
+  "timezone": "",
+  "title": "NetApp Detail: Power",
```
dashboard name need to be added in dashboard_test.go

Hardikl commented on grafana/dashboards/cmode/harvest_dashboard_power.json:
```diff
@@ -0,0 +1,2105 @@
+        "useTags": false
+      },
+      {
+        "allValue": null,
+        "current": {
+          "selected": true,
+          "text": "F2240-127-26_28",
+          "value": "F2240-127-26_28"
+        },
```
need to change

Hardikl commented on grafana/dashboards/cmode/harvest_dashboard_power.json:
```diff
@@ -0,0 +1,2105 @@
+        "type": "datasource"
+      },
+      {
+        "allValue": null,
+        "current": {
+          "selected": false,
+          "text": "Zapi",
+          "value": "Zapi"
+        },
```
same

Hardikl commented on grafana/dashboards/cmode/harvest_dashboard_power.json:
```diff
@@ -0,0 +1,2105 @@
+{
+  "annotations": {
+    "list": [
+      {
+        "builtIn": 1,
+        "datasource": "-- Grafana --",
+        "enable": true,
+        "hide": true,
+        "iconColor": "rgba(0, 211, 255, 1)",
+        "name": "Annotations & Alerts",
+        "target": {
+          "limit": 100,
+          "matchAny": false,
+          "tags": [],
+          "type": "dashboard"
+        },
```
we need to add this block  
```
 "__inputs": [
    {
      "name": "DS_PROMETHEUS",
      "label": "Prometheus",
      "description": "",
      "type": "datasource",
      "pluginId": "prometheus",
      "pluginName": "Prometheus"
    }
  ],
  "__requires": [
    {
      "type": "grafana",
      "id": "grafana",
      "name": "Grafana",
      "version": "7.5.4"
    },
    {
      "type": "panel",
      "id": "graph",
      "name": "Graph",
      "version": ""
    },
    {
      "type": "datasource",
      "id": "prometheus",
      "name": "Prometheus",
      "version": "1.0.0"
    },
    {
      "type": "panel",
      "id": "stat",
      "name": "Stat",
      "version": ""
    },
    {
      "type": "panel",
      "id": "table",
      "name": "Table",
      "version": ""
    }
  ],
```

rahulguptajss commented on cmd/collectors/zapi/plugins/sensor/sensor.go:
```diff
@@ -81,7 +84,12 @@ func (my *Sensor) calculateEnvironmentMetrics(data *matrix.Matrix) ([]*matrix.Ma
 
 	for k, instance := range data.GetInstances() {
-		iKey, iKey2, _ := strings.Cut(k, ".")
+		iKey := instance.GetLabel("node")
+		_, iKey2, found := strings.Cut(k, iKey+".")
+		if !found {
+			my.Logger.Warn().Str("key", iKey+".").Msg("missing instance key")
+			continue
+		}
```
prone to break if instance key order changes as per template. Any better ideas?

cgrinds commented:
nice job @rahulguptajss LGTM once the missing columns on the Shelves table are found 😁
